### ISSUE TITLE
Blind people can't play

### ISSUE THREAD

**Issue by kujiu** [online-go/online-go.com#1366]
Goban are in a canvas and screen readers like Orca, NVDA, TalkBack, or VoiceOver can't reproduce the goban on refreshable braille devices or with text-to-speech. Blind people can't do anything with that.

Please add an ASCII or a table representation of Goban for blind people. We need also to interact with keyboard only.

**Comment by flovo** [comment]
What would be a good representation? I found gnugo has an ascii output with black as 'X' and white as 'O'. The last move is encircled in braces. Star-Points are marked with '+' signs.

For keyboard interaction: typing the coordinate to play is probably the way to go.

GNUGo ASCII goban:
```
    A B C D E F G H J K L M N O P Q R S T
 19 . . . . . . . . . . . . . . . . . . . 19
 18 . . . . . . . . . . . . . . . X . . . 18
 17 . . . . O . . . . . . . . . O . . . . 17
 16 . . O + . . . . . + . . . . . + X . . 16
 15 . . . . . . . . . . . . . . . . . . . 15
 14 . . . . . . . . . . . . . . . . . . . 14
 13 . . . . . . . . . . . . . . . . . . . 13
 12 . . . . . . . . . . . . . . . . . . . 12
 11 . . . . . . . . . . . . . . . . . . . 11
 10 . . . + . . . . . + . . . . . + . . . 10
  9 . . . . . . . . . . . . . . . . . . .  9
  8 . . . . . . . . . . . . . . . . . . .  8
  7 . . . . . . . . . . . . . . . . . . .  7
  6 . . . . . . . . . . . . . . . . . X .  6
  5 . . . . . . . . . . . . . . . O . . .  5
  4 . . . + . . . . . + . . . . . O X . .  4
  3 . . X . . . . . . . . . . . . . X . .  3
  2 . . . . . . . . . . . . . .(O). . . .  2
  1 . . . . . . . . . . . . . . . . . . .  1
    A B C D E F G H J K L M N O P Q R S T
```

**Comment by benjaminpjones** [comment]
I am interested in working on this, so I messaged Kujiu directly and they gave me a lot of great feedback on how this should work.  I'll attach the email thread below, but is a summary of the solution Kujiu sent me:

- The Go board should be represented by an SVG with ARIA labels instead of a canvas.
- Users should be able to use the keyboard to navigate
- The aria labels should have the color first, followed by the position (e.g. "White C4", "Black E3", "No C17" ) these labels would be read by a screen reader when the user focuses on an intersection on the board.
- The focused area should be marked with a high-contrast box

Important notes:

- Many blind users have *some* vision, but not enough to navigate the site like a sighted user.  This means that existing visuals would be nice to keep.
- Accessibility should be on by default (in response to my question about whether we could do this [like lichess](https://lichess.org/blog/U5AX_DcAADkAz-L5/accessibility-for-blind-players), with an accessibility option that shows up if a keyboard-only user is detected)
- Recommended reading: [WCAG](https://www.w3.org/TR/2021/WD-wcag-3.0-20210121/), [WAI-ARIA](https://www.w3.org/TR/2021/CR-wai-aria-1.2-20210302/), [EN301549:2019](https://www.etsi.org/deliver/etsi_en/301500_301599/301549/03.01.01_60/en_301549v030101p.pdf).

If we are going to do this as Kujiu suggested (and I think we should try our best as they have gotten feedback from other blind users as well), then this will be a large undertaking.  If anyone is interested in working on this together, def take a look at the email thread, and also reach out to me! (My handle is **benjito** on OGS)

[kujiu-a11y-discuss-emails-redacted.txt](https://github.com/online-go/online-go.com/files/6305161/kujiu-a11y-discuss-emails-redacted.txt)

**Comment by anoek** [comment]
I think the place to start would be to subclass `GobanCore` in the https://github.com/online-go/goban library. See https://github.com/online-go/goban/blob/master/src/GobanCanvas.ts for the canvas rendering implementation.

Not too long ago I tried out a GobanWebGL renderer and got it functional, the performance was worse than canvas though and I learned some devices have a limit on the number of gl contexts you can have on a page, so I threw it away - however that effort should make it fairly feasible for you to make a `GobanSVG` or `GobanText` renderer, then dropping that in place of GobanCanvas should be fairly straight forward when operating in accessibility mode.

Best of luck!

**Comment by benjaminpjones** [comment]
Sweet! Thanks for the additional info!

**Comment by saizai** [comment]
As I mentioned in the linked issue, I'm situationally blind — fully sighted or fully blind depending on lighting. (It's rare.) I use TalkBack & VoiceOver, but haven't tried doing so with OGS 

I have a few comments in addition to the above.

Absolutely foremost, two things:

1. Talk to actual go players with the disabilities you want to make accessible. I can 100% guarantee you that unless you're used to it, there's a lot you don't know you don't know that is fundamental enough to ruin a design.
2. Try actually using the site yourself with a screen reader. I recommend TalkBack (Android) & VoiceOver (OSX/iPhone) as very easy to learn. You won't realize everything, but you'll definitely know a lot more than just trying to code for a device that you never even use, and a lot that you can't reasonably expect a user who isn't also a web developer to tell you.



Navigation support (e.g. for partial sight or mobility impairment) — using screen readers, zooming tools, etc to navigate and interact with the site — is necessary for, but in this particular case quite different from, usability as a standalone format (i.e. how to show the board).

However, most people with such issues are in fact partial, and most of the gains for most people are to be had with good navigation and markup support — even if it's insufficient for reading the board state blind.

Markup would help e.g. a partially sighted player to tell what's on a given point, and what coordinate that is, assuming that they still use their vision to get the overall picture. 

The points should know their individual coordinates; limited field of vision is a common issue, and following coordinate lines from the edge would be difficult.



For reading the board state totally blind, however, I think the most difficult problem is human memory limits and contextual presentation.

Consider for instance a text to speech screen reader that reads an ASCII board. Suppose (incorrectly, but let's pretend) that it reads it to you correctly.

Left to right, to to bottom, you hear a sequence of empty/black/white, and the occasional newline. Could you really understand this? Even if you get to tell it to repeat whatever you want?

I've used screen readers a lot, and I can navigate fully blind in a crowded conference hall, and my answer is "definitely not, short of extremely high skill at this specific task". Assuming the goal is to make it work for a normal person, not Shūsaku, this is too high a bar to entry.


[Braille display](https://www.humanware.com/en-usa/products/blindness/braille_displays) support seems like one strong possibility.

I personally don't own one — I use TTS — but it seems to me that this would be a viable display for a fully blind go player. Nearly all are one line high, but 25 cells wide or more, so this would require "scrolling" the board one row at a time. That provides enough context, enough at a time (i.e. the fingers can feel a whole line at once, rather than hearing one piece at a time). 

But you'd have to talk to an actual blind go player who uses a braille display. I'm not one. I can just say that I'm pretty sure that TTS wouldn't be able to handle it for anyone short of an expert.

… at least, not unless they're replicating it on a physical board specifically designed for blind go players (tactile pieces, magnetic or pegged, tactile lines, braille coordinates, etc). I've never seen such a thing (and I have looked through catalogues of blind accessible game kits etc), but maybe it exists. If yes, then a TTS is fine.

It would probably also need a coordinate based move entry — e.g. type B15 rather than clicking the point. 


But… really, before trying to tackle "is this accessible totally blind", I suggest addressing the much more common situation of someone with e.g. very blurry tunnel vision using a screen reader interface as navigation and input method (but not as gestalt view).

And being able to use the site, yourself, with an actual screen reader. I promise, they're not so scary. 😉

**Comment by BHydden** [comment]
Nah not stale this would be so great to implement hopefully we can get it done one day.

**Comment by rtayek** [comment]
> I think the place to start would be to subclass `GobanCore` in the https://github.com/online-go/goban library. See https://github.com/online-go/goban/blob/master/src/GobanCanvas.ts for the canvas rendering implementation.
> 

Please consider an option to make a sound (beep) when  an atari happens. I have low vision and I miss or hallucinate ataris and cuts.

Relate links [1](https://forums.online-go.com/t/add-an-option-of-audio-cue-for-visually-impaired-players/50296/2) and [2](https://forums.online-go.com/t/is-there-a-way-to-get-a-sound-beep-when-in-atari/50247/38).

**Comment by rtayek** [comment]
also, i was looking at my go program (which beeps) and it occurred to me that deaf people would not hear the beep. please consider some kind of visual indicator for being in atari.

**Comment by anoek** [comment]
As for the settings toggle I reckon somewhere (probably near the top?) of the sound preferences page would be the place: https://github.com/online-go/online-go.com/blob/devel/src/views/Settings/SoundPreferences.tsx

**Comment by anoek** [comment]
https://github.com/online-go/online-go.com/pull/3295 adds keyboard input support which is  a step in the right direction for this request
